File: 918/solution.go
package _918

func maxSubarraySumCircular(nums []int) int {
	n := len(nums)
	preMax := nums[0]
	curMax := nums[0]
	preMin := nums[0]
	curMin := nums[0]
	total := nums[0]
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	min := func(a, b int) int {
		if a > b {
			return b
		}
		return a
	}
	for i := 1; i < n; i++ {
		preMax = max(preMax+nums[i], nums[i])
		curMax = max(curMax, preMax)
		preMin = min(preMin+nums[i], nums[i])
		curMin = min(curMin, preMin)
		total += nums[i]
	}
	if curMax < 0 {
		return curMax
	} else {
		return max(curMax, total-curMin)
	}
}

File: 918/solution.py
from typing import List


class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        n = len(nums)
        preMax, maxRes = nums[0], nums[0]
        preMin, minRes = nums[0], nums[0]
        total = nums[0]
        for i in range(1, n):
            preMax = max(preMax + nums[i], nums[i])
            maxRes = max(maxRes, preMax)
            preMin = min(preMin + nums[i], nums[i])
            minRes = min(minRes, preMin)
            total += nums[i]
        if maxRes < 0:
            return maxRes
        else:
            return max(maxRes, total - minRes)